pkg/envoy/ads/stream.go

```diff
@@ -30,7 +31,11 @@ func (s *Server) StreamAggregatedResources(server xds_discovery.AggregatedDiscov
 		return errors.Wrap(err, "Could not start Aggregated Discovery Service gRPC stream for newly connected Envoy proxy")
 	}
 
-	// TODO(draychev): check for envoy.ErrTooManyConnections; GitHub Issue https://github.com/openservicemesh/osm/issues/2332
+	// If maxEnvoyConnect is enabled i.e. not 0, then check that the number of Envoy connections is less than maxEnvoyConnect
+	if s.maxEnvoyConnect != 0 && s.currEnvoyConnect >= int64(s.maxEnvoyConnect) {
```

Comment:
2 things:
This is not concurrent-safe, and multiple envoys connecting can race and make this incorrect. Multiple proxies connecting at the same time could evaluate/race this code concurrently and make it through, and afterwards the atomic increase will push the value over maxEnvoyConnect. 

A side suggestion, catalog already has a ConnectedProxy map, and the register/unregister funcs. Potentially the number of proxies could be put there in catalog next the the proxy map, and have the register/unregister funcs update it. I think it makes more sense to be part of catalog, if catalog already keeps track of the proxy list.

Reply:
Thanks for the catch, will look into adding this to catalog!

Reply:
@eduser25 if I added mutex.Lock(), like below, would that prevent racing?
```
	s.dataPlaneMutex.Lock()
	s.currDataPlaneConnected += 1
	if s.maxDataPlaneConnections != 0 && s.currDataPlaneConnected > s.maxDataPlaneConnections {
		return errTooManyConnections
	}
	s.dataPlaneMutex.Unlock()
```